src/core/lib/gprpp/sync.h

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Inline implementation of synchronization primitives, preferred over the C
+   public API for C++ code. */
+
+#ifndef GRPC_CORE_LIB_GPRPP_SYNC_H
+#define GRPC_CORE_LIB_GPRPP_SYNC_H
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/atm.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <assert.h>
+#include <stdint.h>
+#include <atomic>
+#include <type_traits>
+
+#include "src/core/lib/gprpp/memory.h"
+
+namespace grpc_core {
+
+class RefCount {
+ public:
+  explicit RefCount(intptr_t init = 1) : val(init) {}
+
+  // Reinitializes the ref-count.
+  void Reset(intptr_t init) { val.store(init, std::memory_order_relaxed); }
+
+  void Ref(intptr_t n = 1) { val.fetch_add(n, std::memory_order_relaxed); }
+
+  void RefNonZero() {
+#ifndef NDEBUG
+    const intptr_t prior = val.fetch_add(1, std::memory_order_relaxed);
+    assert(prior > 0);
+#else
+    Ref();
+#endif
+  }
+
+  bool RefIfNonZero() {
+    intptr_t prior = val.load(std::memory_order_relaxed);
+    do {
+      if (prior == 0) return false;
+    } while (!val.compare_exchange_weak(prior, prior + 1,
+                                        std::memory_order_release,
+                                        std::memory_order_relaxed));
+    return true;
+  }
+
+  // Returns whether there is only one reference tracked in this refcount.
+  bool IsUnique() const { return val.load(std::memory_order_acquire); }
+
+  bool Unref() {
+    const intptr_t prior = val.fetch_sub(1, std::memory_order_acq_rel);
+    GPR_DEBUG_ASSERT(prior > 0);
+    return prior == 1;
+  }
+
+  intptr_t get() const { return val.load(std::memory_order_relaxed); }
+
+ private:
+  std::atomic<intptr_t> val;
+};
+
+template <typename Child, typename Delete = grpc_core::DefaultDelete<Child>>
+class IntrusivelyRefCounted {
```

Comment:
This could also use a long comment explaining the intended use-cases, along with examples.

It looks like the main difference between this and `RefCount` is that this deletes itself when the refcount goes to zero.  So doesn't this basically do the same thing as our existing `RefCounted` class?

Reply:
Thank you for the pointer. I removed this class. 

We cannot always use `RefCounted` because it's a virtual base class and would add complications for PODs. I gave up on this class and will use RefCount directly

Reply:
Can you say more about what the complications are here?

I would really prefer to use `RefCounted` as much as possible.  One of our goals of that API is to avoid manual ref-counting whenever possible; instead, we want to use `RefCountedPtr<>` and have the compiler enforce the ref-counting.  (This is related to the discussion we had separately about wanting to get the `ClosureRef` change in, so that we can start passing refs along with closures.)  If we instead directly use `RefCount`, then we're not really any better off in that regard than we are with the current `gpr_refcount` API.